File: 06-sobrecarga-igualdad/python/fraccion.py
"""
fraccion.py

EQUIVALENCIAS CON JAVA
  toString()          ->  __str__      (lo que ve el usuario)
                          __repr__     (lo que ve el programador, en la consola)
  equals(Object o)    ->  __eq__(self, otro)
  hashCode()          ->  __hash__

Los metodos con doble guion bajo se llaman "dunder" (double underscore).
No los llamas tu: los llama Python cuando usas print(), == o +.
"""


class Fraccion:

    def __init__(self, numerador, denominador):
        if denominador == 0:
            raise ValueError("El denominador no puede ser 0")
        self._numerador = numerador
        self._denominador = denominador

    # ---------- El equivalente de toString() ----------
    def __str__(self):
        return f"{self._numerador}/{self._denominador}"

    def __repr__(self):
        return f"Fraccion({self._numerador}, {self._denominador})"

    # ---------- "SOBRECARGA" a la Python ----------
    # Python NO permite dos metodos con el mismo nombre: el segundo pisa al primero.
    # El equivalente es UN metodo que revisa el tipo del argumento.
    def sumar(self, otro):
        if isinstance(otro, Fraccion):
            n = self._numerador * otro._denominador + otro._numerador * self._denominador
            d = self._denominador * otro._denominador
            return Fraccion(n, d)
        if isinstance(otro, int):
            return Fraccion(self._numerador + otro * self._denominador, self._denominador)
        raise TypeError(f"No se puede sumar una Fraccion con {type(otro).__name__}")

    # ---------- SOBRECARGA DE OPERADORES: esto Java no lo tiene ----------
    def __add__(self, otro):
        return self.sumar(otro)      # ahora "a + b" funciona

    # ---------- El equivalente de equals() ----------
    def __eq__(self, otro):
        if not isinstance(otro, Fraccion):
            return NotImplemented
        return self._numerador * otro._denominador == otro._numerador * self._denominador

    # Igual que en Java: si defines __eq__, define __hash__
    def __hash__(self):
        return hash(self._numerador / self._denominador)

File: 06-sobrecarga-igualdad/java/Fraccion.java
/*
 * Fraccion.java
 *
 * Esta clase responde las tres preguntas que todo alumno hace:
 *   1) ¿Por que al imprimir mi objeto sale "Fraccion@1b6d3586"?   -> toString()
 *   2) ¿Por que dos objetos iguales dan false con "=="?           -> equals()
 *   3) ¿Puedo tener dos metodos con el mismo nombre?              -> sobrecarga
 */
public class Fraccion {

    private int numerador;
    private int denominador;

    public Fraccion(int numerador, int denominador) {
        if (denominador == 0) {
            throw new IllegalArgumentException("El denominador no puede ser 0");
        }
        this.numerador = numerador;
        this.denominador = denominador;
    }

    // ---------- SOBRECARGA: dos metodos "sumar" distintos ----------
    // Mismo nombre, distinta LISTA DE PARAMETROS. Eso es sobrecarga.
    public Fraccion sumar(Fraccion otra) {
        int n = this.numerador * otra.denominador + otra.numerador * this.denominador;
        int d = this.denominador * otra.denominador;
        return new Fraccion(n, d);
    }

    public Fraccion sumar(int entero) {
        return new Fraccion(this.numerador + entero * this.denominador,
                            this.denominador);
    }

    // ---------- toString(): como se representa el objeto como texto ----------
    // Sin esto, System.out.println(objeto) imprime la direccion de memoria.
    @Override
    public String toString() {
        return this.numerador + "/" + this.denominador;
    }

    // ---------- equals(): que significa "iguales" para ESTA clase ----------
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;                  // es el mismo objeto
        if (obj == null) return false;
        if (!(obj instanceof Fraccion)) return false;  // ni siquiera es una Fraccion

        Fraccion otra = (Fraccion) obj;
        // 1/2 y 2/4 son la misma fraccion: comparamos en cruz
        // (podemos leer otra.numerador aunque sea private: private es POR CLASE)
        return this.numerador * otra.denominador == otra.numerador * this.denominador;
    }

    // Regla de oro de Java: si sobrescribes equals(), sobrescribe hashCode().
    @Override
    public int hashCode() {
        double valor = (double) numerador / denominador;
        return Double.hashCode(valor);
    }
}
